### PR TITLE
Refactor Damage to use Protoypes 

### LINKED ISSUE #3882: Refactor damageable component
TODO:
- [x] ~do not research windowdamage code~
- [x] Refactor damageablecomponent to use prototype damagetypes instead of enum damagetypes
- [x] Refactor damagesystem spaghetti into damageclass prototypes 
- [x] Refactor damagetrigger to use above prototypes
- [x] rename damageclass to damagegroup
- [x] remove DamageFlags in favor of changing resistancesets for godmode™

once this is refactored:
- [x] DrSmugleaf will be freed from his recursion shackles and finish surgery

### PR DESCRIPTION
**Pull request by Silvertorch5**
Closes #3882

### CODE REVIEW

File: Content.Server/Damage/Components/DamageOnHighSpeedImpactComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Damage/Components/DamageOnHighSpeedImpactComponent.cs:27]: This also wont resolve properly without prototype references.

File: Content.Server/Damage/Components/DamageOnLandComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Damage/Components/DamageOnLandComponent.cs:15]: Paul where are the prototype references

File: Content.Server/Damage/Components/DamageOnToolInteractComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Damage/Components/DamageOnToolInteractComponent.cs:28]: Same problem of trying to resolve prototypes directly without prototype references.

File: Content.Server/Damage/Components/DamageOtherOnHitComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Damage/Components/DamageOtherOnHitComponent.cs:27]: This needs to be replaced with resolving the damage type through the prototype manager.

File: Content.Server/Destructible/Thresholds/Triggers/DamageClassTrigger.cs
Comment (DrSmugleaf) [inline Content.Server/Destructible/Thresholds/Triggers/DamageClassTrigger.cs:21]: Another case of needing prototype references for this to work.

File: Content.Server/Destructible/Thresholds/Triggers/DamageTypeTrigger.cs
Comment (DrSmugleaf) [inline Content.Server/Destructible/Thresholds/Triggers/DamageTypeTrigger.cs:30]: Need a direct call to the prototype manager here too, instead of damageable.GetDamageType

File: Content.Server/Doors/Components/ServerDoorComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Doors/Components/ServerDoorComponent.cs:47]: Prototype references

File: Content.Server/GameTicking/Presets/GamePreset.cs
Comment (DrSmugleaf) [inline Content.Server/GameTicking/Presets/GamePreset.cs:71]: Every day we descend one level deeper into potato.
This should be solved with a mob state component refactor probably.

File: Content.Server/GameTicking/Presets/PresetTraitorDeathMatch.cs
Comment (DrSmugleaf) [inline Content.Server/GameTicking/Presets/PresetTraitorDeathMatch.cs:196]: Same goes for this.

File: Content.Server/Light/Components/PoweredLightComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Light/Components/PoweredLightComponent.cs:56]: Prototype references.

File: Content.Server/Medical/Components/HealingComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Medical/Components/HealingComponent.cs:54]: This needs to be a prototype manager index call instead.

File: Content.Server/Medical/Components/MedicalScannerComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Medical/Components/MedicalScannerComponent.cs:104]: These are sent over the network so they should probably be strings instead.

File: Content.Server/Medical/Components/MedicalScannerComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Medical/Components/MedicalScannerComponent.cs:126]: Same applies here.

File: Content.Server/Metabolism/MetabolismComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Metabolism/MetabolismComponent.cs:356]: Prototype manager index call instead.

File: Content.Server/Metabolism/MetabolismComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Metabolism/MetabolismComponent.cs:365]: Prototype manager index call instead.

File: Content.Server/Mining/Components/AsteroidRockComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Mining/Components/AsteroidRockComponent.cs:26]: Prototype references PAUL

File: Content.Server/Nutrition/Components/HungerComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Nutrition/Components/HungerComponent.cs:25]: This isn't your fault but I don't know why this isn't capitalized like baseDecayRate

File: Content.Server/Projectiles/Components/HitscanComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Projectiles/Components/HitscanComponent.cs:60]: The _damage field is redundant, you can make this an auto-property and apply the DataField attribute directly to it.

File: Content.Server/Projectiles/Components/ProjectileComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Projectiles/Components/ProjectileComponent.cs:23]: Prototype references.

File: Content.Server/Projectiles/Components/ProjectileComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Projectiles/Components/ProjectileComponent.cs:26]: Prototype references.

File: Content.Server/Temperature/Components/TemperatureComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Temperature/Components/TemperatureComponent.cs:40]: The _currentTemperature field is also redundant here and can be removed and the DataField attribute moved to the new auto-property.

File: Content.Server/Temperature/Components/TemperatureComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Temperature/Components/TemperatureComponent.cs:60]: These two variables can be moved down if you are moving the rest of the damage dealing code down, they are not needed to be declared this early.

File: Content.Server/Weapon/Melee/MeleeWeaponSystem.cs
Comment (DrSmugleaf) [inline Content.Server/Weapon/Melee/MeleeWeaponSystem.cs:93]: If blunt is intended to be the default then this should be the default of the field for the melee weapon component damage type field, otherwise this should not deal any damage when its null.

File: Content.Server/Weapon/Ranged/ServerRangedWeaponComponent.cs
Comment (DrSmugleaf) [inline Content.Server/Weapon/Ranged/ServerRangedWeaponComponent.cs:171]: These two should be turned into data fields of damage to deal.

File: Content.Shared/Body/Components/SharedBodyComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Body/Components/SharedBodyComponent.cs:200]: This is a tragic side effect of body component and mob state component meeting to create the ultimate display in code that keeps me up at night, but the damage type to deal can at least be moved to a data field to make it less eye grating, along with the amount as well.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:106]: These two methods should be removed and any usages of them replaced with the code inside each.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:140]: Rename the damageGroup argument to group to win instantly.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:157]: Same here.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:261]: Same here.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:408]: I have to say it is less pleasant when easily visible.
The method name should be capitalized though, DamageListToDamageGroup.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:424]: The damage type to deal should be a DataField (that defaults to "Radiation" probably).
I'm sure there are better solutions though.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:438]: Same here, explosion damage should be a DataField.

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Components/DamageableComponent.cs:445]: I don't think prototypes can be directly sent over the network, the string ids should be sent instead and resolved client-side.

File: Content.Shared/Damage/Container/DamageContainerPrototype.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/Container/DamageContainerPrototype.cs:27]: Rename these to _supportedDamageGroupIds and _supportedDamageTypeIds

File: Content.Shared/Damage/DamageGroupPrototype.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/DamageGroupPrototype.cs:27]: This shouldn't be necessary. If it is something is tragically wrong. Either way, dependencies should not be getting resolved inside the AfterDeserialization hook.
```suggestion
```

File: Content.Shared/Damage/DamageSystem.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/DamageSystem.cs:65]: I'm so sorry.

File: Content.Shared/Damage/DamageTypePrototype.cs
Comment (DrSmugleaf) [inline Content.Shared/Damage/DamageTypePrototype.cs:10]: Empty summary.
```suggestion
```

File: Content.Shared/MedicalScanner/SharedMedicalScannerComponent.cs
Comment (DrSmugleaf) [inline Content.Shared/MedicalScanner/SharedMedicalScannerComponent.cs:21]: Use string ids here instead of sending the prototype, I don't think the technology is quite there yet.

File: Resources/Prototypes/Damage/damage.yml
Comment (DrSmugleaf) [inline Resources/Prototypes/Damage/damage.yml:1]: This should be solved when using string ids in data fields instead of the prototypes directly (or when prototype references are in).
Relying on the ordering like this while hilarious is going to keep me up at night.

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Content.Shared/Damage/DamageTypePrototype.cs
Comment (Silvertorch5) [reply Content.Shared/Damage/DamageTypePrototype.cs:10]: this is a valid summary for damage type. it encompasses the void that was before and after

File: Content.Shared/Damage/DamageSystem.cs
Comment (Ygg01) [reply Content.Shared/Damage/DamageSystem.cs:65]: Why are you sorry Dr. PHD Smugleaf?

**Comment by leonsfriedrich** [comment]
Is there a specifc reason no changes have been made to this PR for a while? Because I feel like I could make a lot of the requested changes and have aready started making some on my fork of this branch.  So far I've just added prototype manager index calls, under the assumption that protype references are but a distant dream.

**Comment by ShadowCommander** [comment]
@leonsfriedrich There isn't really a reason. Just people being busy with other things. Feel free to help out with it, since it's a big PR.

Also are you on the Discord?

**Comment by leonsfriedrich** [comment]
@ShadowCommander, yeah Im on discord, as "ElectroSR". I should probably change my name here or there. Just before posting the question here DrSmugleaf responded to a similar question on discord, apparently -Y- already started working on this.

**Comment by leonsfriedrich** [comment]
-Y- was apparently busy and didn't end up doing very much, so I've taken the time to try make most of the changes DrSmugleaf suggested. The only thing left is their first comment: We need to decide whether to add test damage prototypes and a test damage container. So far I haven't ever looked at the testing system so I'm not sure how long that will take (or why it's needed).

In the meantime I have two questsions:

Firstly because this branch has not been updated in a while, but I have been merging from the master branch, the files in my branch have been heavily changed. If I do a PR to this branch, would all of those changes from the master branch clutter the PR and make it hard to see what I have changed? I.e., does someone need to update this branch before I do a PR?

Secondly, the pinned issue #4041 from @metalgearsloth requests that no changes be made to guns. However, this damage changes require changes to HitscanComponent and ProjectileComponent. If this is problem, does the pinned issue mean this is effectively on hold?

And just a comment that PR #4395 will conflict with some of the changes here (mostly just the renaming of damageClass -> damageGroup). Because I'd be starting a PR to branch for a PR, #4395 will probably get merged first so thill will require more here.

File: Content.Server/Chat/Commands/SuicideCommand.cs
Comment (Silvertorch5) [reply Content.Server/Chat/Commands/SuicideCommand.cs:49]: .sucide

File: Content.Server/Mining/Components/AsteroidRockComponent.cs
Comment (Silvertorch5) [reply Content.Server/Mining/Components/AsteroidRockComponent.cs:26]: S a v e  u s

File: Content.Shared/Body/Components/SharedBodyComponent.cs
Comment (Silvertorch5) [reply Content.Shared/Body/Components/SharedBodyComponent.cs:199]: when smug bodysystem sus

File: Content.Shared/Damage/Components/DamageableComponent.cs
Comment (Silvertorch5) [reply Content.Shared/Damage/Components/DamageableComponent.cs:139]: me tiny brain no understand (also its been a month can you recap what i dun goofd here)

Comment (Silvertorch5) [reply Content.Shared/Damage/Components/DamageableComponent.cs:139]: oh wait i understand now i am no longer brain-damaged thanks

File: Content.Shared/Damage/DamageSystem.cs
Comment (Silvertorch5) [reply Content.Shared/Damage/DamageSystem.cs:64]: give me two months eve sub and ill give you a four hour crash course on why smug is sorry 

comes with a free tutorial of IEnumerable examples that crash my pc

File: Content.Server/Damage/Components/DamageOnLandComponent.cs
Comment (Silvertorch5) [reply Content.Server/Damage/Components/DamageOnLandComponent.cs:15]: image

**Review by mirrorcult — APPROVED** [review]
needs updating with master & another review probably

**Comment by Silvertorch5** [comment]
@Zumorica PERMISSION GRANTED FOR INSTAMERGE

**Pull request merged**
